src/main/java/org/opensearch/agent/tools/IndexRoutingTool.java

```diff
@@ -0,0 +1,315 @@
+
+    @Getter
+    @Setter
+    private String inferenceModelId;
+
+    private final MLClients mlClients;
+
+    @Setter
+    private String promptTemplate;
+
+    public IndexRoutingTool(
+        Client client,
+        NamedXContentRegistry xContentRegistry,
+        Integer docSize,
+        Integer k,
+        String embeddingModelId,
+        String inferenceModelId
+    ) {
+        super(
+            client,
+            xContentRegistry,
+            IndexSummaryEmbeddingJob.INDEX_SUMMARY_EMBEDDING_INDEX,
+            IndexSummaryEmbeddingJob.INDEX_SUMMARY_EMBEDDING_FIELD_PREFIX + "_" + embeddingModelId,
+            new String[] {
+                IndexSummaryEmbeddingJob.INDEX_NAME_FIELD,
+                IndexSummaryEmbeddingJob.INDEX_SUMMARY_FIELD,
+                IndexSummaryEmbeddingJob.INDEX_PATTERNS_FIELD,
+                IndexSummaryEmbeddingJob.DATA_STREAM_FIELD,
+                IndexSummaryEmbeddingJob.ALIAS_FIELD },
+            Optional.ofNullable(docSize).orElse(DEFAULT_K),
+            embeddingModelId,
+            Optional.ofNullable(k).orElse(DEFAULT_K)
+        );
+        this.mlClients = new MLClients(client);
+        this.inferenceModelId = inferenceModelId;
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    @Override
+    protected Parser<SearchResponse, Object> searchResponseParser() {
+        return searchResponse -> {
+            SearchHit[] hits = searchResponse.getHits().getHits();
+
+            List<Map<String, Object>> indexSummaries = new ArrayList<>();
+            if (hits != null) {
+                for (SearchHit hit : hits) {
+                    indexSummaries.add(hit.getSourceAsMap());
+                }
+            }
+            return indexSummaries;
+        };
+    }
+
+    @Override
+    public <T> void run(Map<String, String> parameters, ActionListener<T> listener) {
+        // get index of knn-index
+        super.run(parameters, ActionListener.wrap(res -> {
+            List<Map<String, Object>> summaries = (List<Map<String, Object>>) res;
+
+            if (summaries.isEmpty()) {
+                listener.onResponse((T) "No index found");
+                return;
+            }
+
+            // build prompt
+            String summaryStr = summaries
+                .stream()
+                .map(
+                    summaryMap -> String
+                        .format(
+                            Locale.ROOT,
+                            "%s:%s",
+                            summaryMap.get(IndexSummaryEmbeddingJob.INDEX_NAME_FIELD),
+                            summaryMap.get(IndexSummaryEmbeddingJob.INDEX_SUMMARY_FIELD)
+                        )
+                )
+                .collect(Collectors.joining("\\n---\\n"));
+
+            // call LLM, MLModelTool
+            String question = parameters.get(INPUT_FIELD);
+            String prompt = buildPrompt(summaryStr, question);
```

Comment:
We might want to guard the number of tokens in the `summaryStr`, when it is too long, the performance of Claude-instant drops significantly and it is expensive. (I have some idea/implementation in Python, we can discuss how to best proceed there.)

Reply:
right, that's in todo list. let's have a discuss about how to implement it.